src/condor_utils/proc_family_direct_cgroup_v2.cpp

```diff
@@ -350,9 +365,189 @@ ProcFamilyDirectCgroupV2::cgroupify_process(const std::string &cgroup_name, pid_
+	// RET # (emit this block for every blocked major/minor pair)<--+
+	// good-return-label:
+	// RET
+	//
+
+	// R0 = 1; (i.e. good return)
+	//                             reg/imm
+	insn.code = BPF_MOV | BPF_K  | BPF_ALU;
+	insn.dst_reg = BPF_REG_0;
+	insn.src_reg = 0;
+	insn.off     = 0;
+	insn.imm     = 1;
+	bpf_program.emplace_back(insn);
+
+	// R2 = *(R1 + 4) # R2 = ctx->device_major
+	insn.code = BPF_LDX | BPF_MEM; 
+	insn.dst_reg = BPF_REG_2;
+	insn.src_reg = BPF_REG_1;
+	insn.off     = 4;
+	insn.imm     = 0;
+	bpf_program.emplace_back(insn);
+
+	// R3 = *(R1 + 8) # R3 = ctx->device minor
+	insn.code = BPF_LDX | BPF_MEM;
+	insn.dst_reg = BPF_REG_3;
+	insn.src_reg = BPF_REG_1;
+	insn.off     = 8;
+	insn.imm     = 0;
+	bpf_program.emplace_back(insn);
+
+	for (const dev_t major_and_minor: this->cgroup_hide_devices) {
+		// JNE if R2 != major_device, PC += 3 (units of jmp offset are insns, not bytes)
+		insn.code = BPF_JMP32 | BPF_JNE | BPF_K;
+		insn.dst_reg = BPF_REG_2;
+		insn.src_reg = 0;
+		insn.off     = 3;
+		insn.imm     = major(major_and_minor);
+		bpf_program.emplace_back(insn);
+
+		// JNE if R3 != minor, PC += 2 (units of jmp offset are insns, not bytes)
+		insn.code = BPF_JMP32 | BPF_JNE | BPF_K;
+		insn.dst_reg = BPF_REG_3;
+		insn.src_reg = 0;
+		insn.off     = 2;
+		insn.imm     = minor(major_and_minor);
+		bpf_program.emplace_back(insn);
+
+		// R0 = 0; (i.e. fail, as we are fixing to return)
+		//                             reg/imm
+		insn.code = BPF_MOV | BPF_K  | BPF_ALU;
+		insn.dst_reg = BPF_REG_0;
+		insn.src_reg = 0;
+		insn.off     = 0;
+		insn.imm     = 0;
+		bpf_program.emplace_back(insn);
+
+		// RET -- RET with R0 = 0 (Assigned above)
+		//                    i.e. fail with EPERM
+		insn.code = BPF_JMP | BPF_EXIT | BPF_K;
+		insn.dst_reg = BPF_REG_0;
+		insn.src_reg = BPF_REG_0;
+		insn.off     = 0;
+		insn.imm     = 0;
+		bpf_program.emplace_back(insn);
+	}
+
+	// RET -- RET with R0 = 0 means block this access with EPERM, no matter what file perms say
+	//            with R0 = 1 means return successfully, but still check file perms
+	insn.code = BPF_JMP | BPF_EXIT | BPF_K;
+	insn.dst_reg = BPF_REG_0;
+	insn.src_reg = BPF_REG_0;
+	insn.off     = 0;
+	insn.imm     = 0;
+	bpf_program.emplace_back(insn);
+
+	const char *bpf_license = "Apache 2.0"; // Just like the rest of HTCondor
+	char bpf_log[512];
+	memset(bpf_log, '\0', sizeof(bpf_log));
+
+	union bpf_attr attr;
+	memset(&attr, '\0', sizeof(attr)); // Avoid no end of warnings
+
+	attr.prog_type = BPF_PROG_TYPE_CGROUP_DEVICE;
+	attr.insns     = (__u64) bpf_program.data();
+	attr.insn_cnt  = bpf_program.size();
+	attr.license   = (__u64) bpf_license;
+
+	// Setting up the log causes the bpf system call to fail with ENOSPC
+	// even when it successfully writes a useful log message to the buffer.
+	// So, try to load once without the log, if that succeeds, go on.
+	// if it fails, try again with the log configured, so we can see the error
```

Comment:
This seems moderately terrifying.  Do you understand why?

Reply:
Could not figure it out, after hours of hacking.  Even went looking into the kernel source code (which was helpful for understanding other BPF instructions...)  Other systems that install bpf program seem to do the same things.